pkg/controller/admissionchecks/provisioning/controller.go

```diff
@@ -122,56 +122,69 @@ func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reco
+		return reconcile.Result{}, err
+	}
+
 	// get the lists of relevant checks
 	relevantChecks, err := admissioncheck.FilterForController(ctx, c.client, wl.Status.AdmissionChecks, kueue.ProvisioningRequestControllerName)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	list := &autoscaling.ProvisioningRequestList{}
-	if err := c.client.List(ctx, list, client.InNamespace(wl.Namespace), client.MatchingFields{RequestsOwnedByWorkloadKey: wl.Name}); client.IgnoreNotFound(err) != nil {
-		return reconcile.Result{}, err
+	checkConfig := make(map[string]*kueue.ProvisioningRequestConfig, len(relevantChecks))
+	for _, checkName := range relevantChecks {
+		prc, err := c.helper.ConfigForAdmissionCheck(ctx, checkName)
+		if client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+		checkConfig[checkName] = prc
 	}
-	ownedPrs := list.Items
-	activeOrLastPRForChecks := c.activeOrLastPRForChecks(ctx, wl, relevantChecks, ownedPrs)
+
+	activeOrLastPRForChecks := c.activeOrLastPRForChecks(ctx, wl, checkConfig, list.Items)
 
 	wlInfo := workloadInfo{
 		checkStates: make([]kueue.AdmissionCheckState, 0),
 	}
-	err = c.syncCheckStates(ctx, wl, &wlInfo, relevantChecks, activeOrLastPRForChecks)
+	err = c.syncCheckStates(ctx, wl, &wlInfo, checkConfig, activeOrLastPRForChecks)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	err = c.deleteUnusedProvisioningRequests(ctx, ownedPrs, activeOrLastPRForChecks)
+	err = c.deleteUnusedProvisioningRequests(ctx, list.Items, activeOrLastPRForChecks)
 	if err != nil {
 		log.V(2).Error(err, "syncOwnedProvisionRequest failed to delete unused provisioning requests")
 		return reconcile.Result{}, err
 	}
 
-	requeAfter, err := c.syncOwnedProvisionRequest(ctx, wl, &wlInfo, relevantChecks, activeOrLastPRForChecks)
+	requeueAfter, err := c.syncOwnedProvisionRequest(ctx, wl, &wlInfo, checkConfig, activeOrLastPRForChecks)
 	if err != nil {
 		// this can also delete unneeded checks
 		log.V(2).Error(err, "syncOwnedProvisionRequest failed")
 		return reconcile.Result{}, err
 	}
 
-	if requeAfter != nil {
-		return reconcile.Result{RequeueAfter: *requeAfter}, nil
+	if requeueAfter != nil {
+		return reconcile.Result{RequeueAfter: *requeueAfter}, nil
 	}
 	return reconcile.Result{}, nil
 }
 
-func (c *Controller) activeOrLastPRForChecks(ctx context.Context, wl *kueue.Workload, relevantChecks []string, ownedPrs []autoscaling.ProvisioningRequest) map[string]*autoscaling.ProvisioningRequest {
+func (c *Controller) activeOrLastPRForChecks(
+	ctx context.Context,
+	wl *kueue.Workload,
+	checkConfig map[string]*kueue.ProvisioningRequestConfig,
+	ownedPrs []autoscaling.ProvisioningRequest,
+) map[string]*autoscaling.ProvisioningRequest {
 	activeOrLastPRForChecks := make(map[string]*autoscaling.ProvisioningRequest)
 	log := ctrl.LoggerFrom(ctx)
-	for _, checkName := range relevantChecks {
+	for checkName, prc := range checkConfig {
 		for i := range ownedPrs {
```

Comment:
Could we move this logic to a new function e.g. `getProvReqForCheck` that would return ProvReq which matches AdmissionCheck?

Reply:
In that case we need to return list of PRs, correct?

Reply:
Because it's failing on https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/admissionchecks/provisioning/controller_test.go#L1396-L1404.